iree/compiler/Dialect/Flow/Transforms/HLOToHLOPreprocessing.cpp

```diff
@@ -307,6 +307,318 @@ class AdjustDepthwiseFilterShape : public OpRewritePattern<mhlo::ConvOp> {
+                                      .int_value_begin())
+                                    .getLimitedValue();
+    if (lhsContractionDim == 2 && rhsContractionDim == 1) return failure();
+
+    Value lhs = op.lhs(), rhs = op.rhs();
+
+    auto make1DElementsAttr = [&rewriter](ArrayRef<int64_t> integers) {
+      auto type = RankedTensorType::get({static_cast<int64_t>(integers.size())},
+                                        rewriter.getIntegerType(64));
+      return DenseIntElementsAttr::get(type, integers);
+    };
+
+    // Transpose lhs input.
+    if (lhsContractionDim == 1) {
+      Type transposedType = RankedTensorType::get(
+          {lhsShapeType.getDimSize(0), lhsShapeType.getDimSize(2),
+           lhsShapeType.getDimSize(1)},
+          resultType.getElementType());
+      lhs = rewriter.create<mhlo::TransposeOp>(op.getLoc(), transposedType, lhs,
+                                               make1DElementsAttr({0, 2, 1}));
+    }
+
+    if (rhsContractionDim == 2) {
+      Type transposedType = RankedTensorType::get(
+          {rhsShapeType.getDimSize(0), rhsShapeType.getDimSize(2),
+           rhsShapeType.getDimSize(1)},
+          resultType.getElementType());
+      rhs = rewriter.create<mhlo::TransposeOp>(op.getLoc(), transposedType, rhs,
+                                               make1DElementsAttr({0, 2, 1}));
+    }
+
+    auto dimensionNumbers = mhlo::DotDimensionNumbers::get(
+        /*lhs_batching_dimensions=*/make1DElementsAttr({0}),
+        /*rhs_batching_dimensions=*/make1DElementsAttr({0}),
+        /*lhs_contracting_dimensions=*/make1DElementsAttr({2}),
+        /*rhs_contracting_dimensions=*/
+        make1DElementsAttr({1}), rewriter.getContext());
+
+    Value result = rewriter.create<mhlo::DotGeneralOp>(
+        op.getLoc(), op.getType(), lhs, rhs, dimensionNumbers,
+        op.precision_config().hasValue() ? op.precision_config().getValue()
+                                         : nullptr);
+    rewriter.replaceOp(op, result);
+    return success();
+  }
+};
+
+// Rewrite mhlo.dot_general to operate on rank-3 tensors when reduction dims are
+// in consequative order and not spliting the domain.
+class RankReducedDotGeneral : public OpRewritePattern<mhlo::DotGeneralOp> {
+ public:
+  using OpRewritePattern<mhlo::DotGeneralOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mhlo::DotGeneralOp op,
+                                PatternRewriter &rewriter) const override {
+    auto lhsShapeType = op.lhs().getType().cast<ShapedType>();
+    auto rhsShapeType = op.rhs().getType().cast<ShapedType>();
+    auto resultType = op.getResult().getType().cast<ShapedType>();
+
+    if (!lhsShapeType || !rhsShapeType || !resultType) return failure();
+    if (!lhsShapeType.hasStaticShape() || !rhsShapeType.hasStaticShape())
+      return failure();
+    if (resultType.getRank() <= 3) return failure();
+
+    auto extract1DVector = [](DenseIntElementsAttr elements) {
```

Comment:
This is defined multiple times. Maybe just make a utility method. Alternative is to replace extract1DVector with

```
llvm::to_vector<6>(llvm::make_range(elements, [](APInt v){ return v.getSExtValue(); }))
```